src/Nethermind/Nethermind.Evm/Tracing/EstimateGasTracer.cs

```diff
@@ -194,10 +197,73 @@ internal long CalculateAdditionalGasRequired(Transaction tx, IReleaseSpec releas
             return _currentGasAndNesting.Peek().AdditionalGasRequired + RefundHelper.CalculateClaimableRefund(intrinsicGas + NonIntrinsicGasSpentBeforeRefund, TotalRefund, releaseSpec);
         }
 
-        public long CalculateEstimate(Transaction tx, IReleaseSpec releaseSpec)
+        private static bool Executable(Transaction transaction, BlockHeader block, GethLikeTxTracer txTracer, UInt256 gas, ITransactionProcessor transactionProcessor)
         {
+            transaction.GasLimit = (long)gas;
+            
+            try
+            {
+                transactionProcessor.CallAndRestore(transaction, block, txTracer);
+                string gethTracer = new EthereumJsonSerializer().Serialize(txTracer.BuildResult(), true);
+
+                return !(gethTracer.Contains("\"failed\": true") || gethTracer.Contains("OutOfGas"));
+            }
+            catch (Exception)
+            {
+                return false;
+            }
+        }
+
+        private long BinarySearchEstimate(UInt256 leftBound, UInt256 rightBound, UInt256 cap, Transaction tx, BlockHeader header, ITransactionProcessor transactionProcessor)
+        {
+            while (leftBound + 1 < rightBound)
+            {
+                UInt256 mid = (leftBound + rightBound) / 2;
+                if (!Executable(tx, header, new GethLikeTxTracer(GethTraceOptions.Default), mid, transactionProcessor))
```

Comment:
Isn't CallOutputTracer enough? Also it should be created where its output is used, so probably in the Executable method